source/docs/design/returning-mutable-references.md

```diff
@@ -0,0 +1,294 @@
+  // mutable borrow 'a expires here
+  // error: requires condition after the expiry of the borrow
+  sort(&'b mut v); // ensures elements are the same
+  assert(sum(v) >= 20);
+  assert(exists(|i: nat| i < v.len() && v.index(i) == v_prev.index(i) + v2.index(4)));
+}
+```
+
+Example 3. Error, `after` `ensures` only takes hold at the end of the borrow:
+
+```rust
+fn caller(v: Vec<u64>, v2: Vec<u64>) {
+  requires([sum(v) >= 20, v.len() > 10, v2.len() == v1.len()]);
+  #[spec] let v_prev = v;
+  let e: &'a mut u64 = v.index_mut_increasing(4);
+  e = e + v2.index(4);
+  assert(sum(v) > 20); // error: v still has the old value (before the borrow)
+                       // emit warning that there's a live mutable borrow?
+  // mutable borrow 'a expires here
+  sort(&'b mut v); // ensures elements are the same
+  assert(exists(|i: nat| i < v.len() && v.index(i) == v_prev.index(i) + v2.index(4)));
+}
+```
+
+Example 4. Example 3 fixed: explicitly drop the borrowing variable to terminate the borrow:
+
+```rust
+fn caller(v: Vec<u64>, v2: Vec<u64>) {
+  requires([sum(v) >= 20, v.len() > 10, v2.len() == v1.len()]);
+  #[spec] let v_prev = v;
+  let e: &'a mut u64 = v.index_mut_increasing(4);
+  e = e + v2.index(4);
+  expire(e); // force the end of the borrow here
+  assert(sum(v) > 20); // now verifies
+  sort(&'b mut v); // ensures elements are the same
+  assert(exists(|i: nat| i < v.len() && v.index(i) == v_prev.index(i) + v2.index(4)));
+}
+```
+
+This discussed the general case of a borrow extending for potentially multiple statements. The special case of a single direct assignment to the mutable reference in the same statement as the function call can be handled using the same semantics as described above (but without the need for an explicit `expire`).
+
+Example 5.
+
+```rust
+fn caller(v: Vec<u64>, v2: Vec<u64>) {
+  requires([sum(v) >= 20, v.len() > 10, v2.len() == v1.len()]);
+  let e = *v.index(4);
+  #[spec] let v_prev = v;
+  *v.index_mut_increasing(4) = e + v2.index(4); // single statement with direct assigment
+  assert(sum(v) > 20);
+  sort(&'b mut v);
+  assert(exists(|i: nat| i < v.len() && v.index(i) == v_prev.index(i) + v2.index(4)));
+}
+```
+
+Example 6. `Option<&mut>` brought up by @tjhance. A function on a hashtable to optionally obtain a mutable reference to a value, if the key is present:
+
+```rust
+impl HashTable<K, V> {
+  pub fn borrow_mut<'a>(&'a mut self, key: K) -> Option<&'a mut V> {
+    ensures(|ref: &mut V| if self.contains(key) {
+      ref.is_some() && ref.value() == old(self).get(key)
+    } else {
+      equal(*ref, None)
+    }));
+    after<'a>(|ref: &mut V| {
+      ensures([
+        // forall(|k: K| equal(self.contains(k), old(self).contains(k)) && (self.contains(k) >>= (
+        // 	 self.get(k) == (if k == key { ref.value() } else { old(self).get(k) })
+        // ))),
+        self.view() == old(self).view().set(key, *ref.value()),
+      ])
+    })
+    
+    // ...
+  }
+}
```

Comment:
I think there is something missing in this example, or I don't understand it. How is ref: V in the after clause tied to the field of the 'Some' constructor in the Option?